torch_ttnn/passes/memory_pass.py

```diff
@@ -0,0 +1,240 @@
+                        is_follow_up_tt_compute = True
+                    if user.target in set(
+                        [ttnn.from_device, ttnn.to_torch, ttnn.to_layout]
+                    ):
+                        is_follow_up_from_device = True
+
+                # If no follow up ttnn op using the output of current node
+                # and if one of the users is a from device operation,
+                # then swap out output tensor from device
+                if not is_follow_up_tt_compute and is_follow_up_from_device:
+                    mm.used_sram -= output_size
+                    mm.free_sram += output_size
+                    mm.tensors_on_device.remove(tid)
+
+                    on_device_meta.remove((tid, output_size))
+
+                    mm.logs += f"op removed from device: {node.name}\n"
+                    mm.logs += f"output size: {output_size} bytes\n"
+                    mm.logs += f"total SRAM usage: {mm.used_sram} bytes\n"
+
+                    print(f"op removed from device: {node.name}")
+                    print(f"output size: {output_size} bytes")
+                    print(f"total SRAM usage: {mm.used_sram} bytes")
+                # Once current node's input & output tensor processing is over, mark it as computed
+                computed_nodes.append(node)
+                # Swap out input tensors from device if its usage is completed
+                # i.e there are no ttnn ops on device which uses the tensor
+                for input_node in node.all_input_nodes:
+                    # Exception: Shape node is just like a parameter containing shape of input tensor
+                    if input_node.target == ttnn.Shape:
+                        continue
+                    keep_on_device = False
+                    input_node_users = list(input_node.users.keys())
+                    for user in input_node_users:
+                        if user not in computed_nodes and is_tt_compute(user):
+                            keep_on_device = True
+                    if keep_on_device is False:
+                        input_size = op_registry.get_size(
+                            op_registry.get_shape(input_node),
+                            op_registry.get_dtype(input_node),
+                        )
+                        mm.used_sram -= input_size
+                        mm.free_sram += input_size
+                        tid = mm.node_to_tid_map[input_node]
+                        mm.tensors_on_device.remove(tid)
+                        on_device_meta.remove((tid, input_size))
+
+                        mm.logs += (
+                            f"input tensor removed from device: {input_size} bytes\n"
+                        )
+                        mm.logs += f"total SRAM usage: {mm.used_sram} bytes\n"
+
+                        print(f"input tensor removed from device: {input_size} bytes")
+                        print(f"total SRAM usage: {mm.used_sram} bytes")
```

Comment:
Indentation level here is huge `                        `, not ok. So many signals to break the thing into pieces